src/test/consensus/DistributedValidatorsSim_test.cpp

```diff
@@ -0,0 +1,324 @@
+#include <BeastConfig.h>
+#include <ripple/beast/clock/manual_clock.h>
+#include <ripple/beast/unit_test.h>
+#include <test/csf.h>
+#include <utility>
+
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/classification.hpp>
+#include <algorithm>
+#include <sstream>
+#include <fstream>
+#include <string>
+
+namespace ripple {
+namespace test {
+
+namespace csf
+{
+
+template<class SizePDF, class Generator>
+void
+randomRankedConnect(
+        PeerGroup &peers,
+        std::vector<double> const &ranks,
+        int numConnections,
+        SizePDF numConnectionsPDF,
+        std::chrono::milliseconds delay,
+        Generator &g)
+{
+    std::size_t const size = peers.size();
+    assert(size == ranks.size());
+    (void)size;
+
+    // 2. Generate UNLs based on sampling without replacement according
+    //    to weights.
+    std::vector<PeerGroup> connectionLists(numConnections);
+    std::vector<Peer *> rawPeers(peers.begin(), peers.end());
+    std::generate(connectionLists.begin(), connectionLists.end(), [&]() {
+        std::vector<Peer *> res =
+                random_weighted_shuffle(rawPeers, ranks, g);
+        res.resize(numConnectionsPDF(g));
+        return PeerGroup(std::move(res));
+    });
+
+    // 3. Assign trust
+    std::uniform_int_distribution<int> u(0, numConnections - 1);
+    for (auto &peer : peers)
+    {
+        for (auto &target : connectionLists[u(g)])
+            peer->connect(*target, delay);
+    }
+}
+
+} // csf
+
+/** In progress simulations for diversifying and distributing validators
+*/
+class DistributedValidators_test : public beast::unit_test::suite
+{
+
+    void
+    completeTrustCompleteConnectFixedDelay(
+            std::size_t numPeers,
+            std::chrono::milliseconds delay = std::chrono::milliseconds(200),
+            bool printHeaders = false)
+    {
+        using namespace csf;
+        using namespace std::chrono;
+
+        // Initialize persistent collector logs specific to this method
+        std::string const prefix =
+                "DistributedValidators_"
+                "completeTrustCompleteConnectFixedDelay";
+        std::fstream
+                txLog(prefix + "_tx.csv", std::ofstream::app),
+                ledgerLog(prefix + "_ledger.csv", std::ofstream::app);
+
+        // title
+        log << prefix << "(" << numPeers << "," << delay.count() << ")"
+            << std::endl;
+
+        // number of peers, UNLs, connections
+        BEAST_EXPECT(numPeers >= 1);
+
+        ConsensusParms parms;
+        Sim sim;
+        PeerGroup peers = sim.createGroup(numPeers);
+
+        // complete trust graph
+        peers.trust(peers);
+
+        // complete connect graph with fixed delay
+        peers.connect(peers, delay);
+
+        // Initialize collectors to track statistics to report
+        TxCollector txCollector;
+        LedgerCollector ledgerCollector;
+        auto colls = collectors(txCollector, ledgerCollector);
```

Comment:
That idiom of having a factory function is used in several spots in the framework.  The preference is to rename it `make_foo` instead of `foo` when it is used to build `Foo`s? My reason for leaning towards `foo` is that once we have class template argument deduction, we can drop the factories and use `Foo` directly.

Reply:
Speaking only for myself, I'd like to see `make_foo` until we reach the point that we're actually _using_ class template argument deduction.  For right now you're not calling the constructor; the `make_` helps to signal that to the reader.  For this code base the `make_` idiom is pervasive, so you'll have good company.

Once class argument deduction works you really will be calling the constructor.  And many of those `make_` functions will disappear from the code base at art time.  I'm not hard over on this, just thinking about my stumble on the initial read...